tools/update.php

```diff
@@ -0,0 +1,747 @@
+    //   [1] => /var/www/loris/SQL/Release_patches/17.0_To_18.0_upgrade_B.sql
+    //   [2] => /var/www/loris/SQL/Release_patches/18.0_To_19.0_upgrade.sql
+    //   [3] => /var/www/loris/SQL/Release_patches/19.0_To_19.1_upgrade.sql
+    // Range will run once when $start = $end. We want this in the case where
+    // there is no difference in major versions but there is a change in minor
+    // versions. The below loops will, in this case, match the minor release
+    // patches.
+    $start = $from_versions[MAJOR];
+    $end   = $start; // default case: there is a minor release difference
+    if ($diff_major > 0) {
+        $end = $to_versions[MAJOR];
+    }
+    $patch_dir           = $loris_root . 'SQL/Release_patches/';
+    $all_release_patches = glob($patch_dir . '*.sql');
+    $patches = [];
+    if ($diff_major > 0 || $diff_minor > 0) {
+        foreach (range($start, $end) as $v) {
+            foreach ($all_release_patches as $patch) {
+                // strpos must be 0 or else this will match w/ e.g. 16.0_To_17.0
+                // when the server is already on version 17
+                if (strpos(basename($patch), "$v") === 0) {
+                    $patches[] = $patch;
+                }
+            }
+        }
+    }
+    if (isDev()) {
+        echo "[+] Developer instance detected. Including developer patches..."
+            . PHP_EOL;
+
+        // Add all patches in Archive/$MAJOR.$MINOR. Everything other needed
+        // command will be in the Release patches which are been added above
+        $dev_patch_dir = $loris_root . 'SQL/Archive/' . $to_versions[MAJOR]
+            . '.' . $to_versions[MINOR] . '/';
+        $patches       = array_merge($patches, glob($dev_patch_dir . '*.sql'));
+    }
+    return $patches;
+}
+
+/**
+ * Apply a list of SQL patch files using exec.  Currently this function is only
+ * called with $report_only = true so patches are never applied. This can be
+ * changed in the future if this functionality is required.
+ *
+ * @param array      $patches       Patch files to be applied
+ * @param dictionary $db_config     DB config info. Needed for credentials
+ * @param bool       $apply_patches If true, MySQL commands will be run
+ *
+ * @return bool Always true (for now). False if patches fail to be applied.
+ */
+function applyPatches($patches, $db_config, $apply_patches = false) : bool
+{
+    // Iterate over all patches and source them into MySQL
+    $A = $db_config['database'];
+    $u = $db_config['username'];
+    $p = $db_config['password'];
+    $h = $db_config['host'];
+    if ($apply_patches === false) {
+        echo '[-] NOTE: Running in Report-Only mode. Patching commands will be '
+            . 'displayed but not executed.' . PHP_EOL;
+        sleep(1);
+    }
+    while (count($patches) > 0) {
+        $patch = array_shift($patches);
+        $cmd   = "mysql -h $h -u $u -p -A $A < $patch";
+        if ($apply_patches === true) {
+            if (doExec($cmd) === false) {
```

Comment:
Does it really apply the patch? No password is passed to the DoExec function and no where the password is passed as a prompt. So the user need to know the DB pwd (and enter it maybe multiple time).

Reply:
You're right, it would probably be best to import the settings from Config so that the user doesn't need to do this manually.

Reply:
Ah nevermind I remember why I did it this way:

1.  The password is not passed to `doExec` because this function always prints out the command. If the password is passed this way then it will appear in plain text in the console which isn't desirable.
2) I think getting the user to enter the password manually is actually a good way to be _really_ sure that the user actually wants to make changes to the DB. While this can be done automatically it's probably good to have this inconvenience in the interest of data integrity.
3) While the user may need to enter the DB password manually, and perhaps multiple times, this is what they would have to do anyway without this script. So it's no worse than the current procedure.